### ISSUE TITLE
Attention modifications

### ISSUE THREAD

**Issue by PrajitR** [tensorflow/tensorflow#640]
In the `attention_decoder` of `seq2seq.py`, attention is computed by the statements 
```python
cell_output, new_state = cell(x, states[-1])
attns = attention(new_state)
with vs.variable_scope("AttnOutputProjection"):
  output = rnn_cell.linear([cell_output] + attns, output_size, True)
``` 
However, in Grammar as a Foreign Language, attention is computed using the last hidden vector of the multi-layer LSTM (e.g. `cell_output`). Computing attention with `new_state` involves using every layer's state, including their cell state (which should probably be only used inside the LSTM layer). To match GaaFL, the attention call should be
```python
attns = attention(cell_output)
```
Of course, if the `cell` is wrapped by an `OutputProjectionWrapper` as done in `embedding_attention_seq2seq`, then this call is computationally expensive. To prevent this, `embedding_attention_seq2seq` can be modified as such:
```python
# Starting from line 606.
output_size = None
if output_projection is None:
  output_size = num_decoder_symbols # Projection wrapper removed.
```
Now, `cell_output` will be the last hidden vector of the multi-layered LSTM. The reason this works is because `output` is already being projected for the softmax, so `cell_output` does not need to be projected to a length of `num_decoder_symbols`. If as before`cell` is wrapped by `OutputProjectionWrapper`, the `cell_output` will have size `num_decoder_symbols`. Computing the `output` will then involve a `(num_decoder_symbols + num_heads * attn_size) x num_decoder_symbols` matrix, which can be huge. However, if we make the change, then the matrix is only of size `(cell.output_size + num_heads * attn_size) x num_decoder_symbols`.

Thoughts? Have I missed something important? If this sounds good, I can make a PR. Thanks!

**Comment by lukaszkaiser** [comment]
Thanks for spotting the bug - you're right that the code uses the whole multi-layer state to compute the attention query while in the paper it was only the output! It should be corrected, but I'm not sure if your solution will work well if the cell is already wrapped in an OutputProjectionWrapper. I mean - it will work technically, but it will basically first apply the cell to the end (to get a vector of size cell.output_size) and then it'll project it back to just output_size (we should change these names too, I'm sorry for the poor choice of variable names in this function).

I see a few possible solutions. The cleanest might be to forbid the use of attention_decoder without output_projection - in fact most use-cases already specify the output projection separately. But, since this is a breaking change, maybe we should start as you suggest. I'll be happy to review a code if you want to write it, of course!

**Comment by PrajitR** [comment]
You're totally right about the case when the cell is already wrapped in an OutputProjectionWrapper. However, in this case also, we should remove the line where the cell is again wrapped in OutputProjectionWrapper -- the hidden state would be projected to cell.output_size (which probably equals num_decoder_symbols), and then reprojected to num_decoder_symbols.

Since forbidding attention_decoder without output_projection would be a breaking change, one thing we can do is to check if `isinstance(cell, OutputProjectionWrapper)` and print a warning telling the user to specify output_projection instead. Then in future versions of TF, a `ValueError` can be returned in this case and output_projection will not have a default value.

**Comment by markusdr** [comment]
This issue probably contributes to the bad translation results people have reported from running the translation demo code (see [https://github.com/tensorflow/tensorflow/issues/550] ). @PrajitR, I understand it may not be straightforward to provide a fix for the general seq2seq case, but what might the fix specifically for the *translate* demo implementation look like?

**Comment by ebrevdo** [comment]
I vote we get rid of OutputProjectionWrapper completely.  This is the kind
of problem that it contributes to.  It's also much slower than doing a
projection on the whole output.

**Comment by lukaszkaiser** [comment]
I'm not so sure about 2 points made above. I agree with the general idea though, but before we accept the statements as facts we need to think about them.

(1) I'm not sure that this issue actually makes the models worse. I'll run experiments next week to check it -- it might well be. But it's not clear - does anyone have numbers to support it? I'll try to get some numbers next week.

(2) I'm not sure using OutputProjectionWrapper is much slower. In fact I ran a few simple tests and it was at most marginally slower or the same. Eugene - do you have numbers for that?

Still - we should think about making the attention and output projection situation cleaner. And I think this will require breaking the current API -- this part was maybe just not designed well enough. Let's think how to make it right.

**Comment by LeavesBreathe** [comment]
>(1) I'm not sure that this issue actually makes the models worse. I'll run experiments next week to check it -- it might well be. But it's not clear - does anyone have numbers to support it? I'll try to get some numbers next week.

I don't have numbers, but I will say that I have used this seq2seq example pretty heavily with the attention on all hidden states from each layer. 

Also, maybe I'm missing something here, but can't we just slice the hidden state before we input it into the attention mechanism? If we are only interested in the hidden state output of the last layer, why don't we just slice that part from the total hidden state?

```python
cell_output, new_state = cell(x, states[-1])

#I need to do some thinking on begin and size args
input_state_attn = tf.slice(newstate, begin, size)
attns = attention(input_state_attn)

with vs.variable_scope("AttnOutputProjection"):
  output = rnn_cell.linear([cell_output] + attns, output_size, True)
```

**Comment by jstaker7** [comment]
I've been curious about slicing the output state as well, as mentioned by LeavesBreathe.

**Comment by peterjliu** [comment]
Any update on a plan to resolve this?

**Comment by girving** [comment]
@lukaszkaiser @ebrevdo Can you one of you modify the title, close, mark as contributions welcome, or similar?  "Attention modifications" isn't very informative as a title.

**Comment by DogNick** [comment]
Is it resolved?  so the code in seq2seq.py has been modified to just make outprojection of the last layer?
where can i see the latest version?
